app/scripts/keyring-controller.js

```diff
@@ -0,0 +1,572 @@
+        const firstAccount = accounts[0]
+        const hexAccount = normalize(firstAccount)
+        this.configManager.setSelectedAccount(hexAccount)
+        this.setupAccounts(accounts)
+
+        this.emit('update')
+        cb(null, this.getState())
+      })
+    })
+  }
+
+  migrateAndGetKey (password) {
+    let key
+    const shouldMigrate = !!this.configManager.getWallet() && !this.configManager.getVault()
+
+    return this.loadKey(password)
+    .then((derivedKey) => {
+      key = derivedKey
+      this.key = key
+      return this.idStoreMigrator.oldSeedForPassword(password)
+    })
+    .then((serialized) => {
+      if (serialized && shouldMigrate) {
+        const keyring = this.restoreKeyring(serialized)
+        this.keyrings.push(keyring)
+        this.configManager.setSelectedAccount(keyring.getAccounts()[0])
+      }
+      return key
+    })
+  }
+
+  createNewVault (password, entropy, cb) {
+    const configManager = this.configManager
+    const salt = this.encryptor.generateSalt()
+    configManager.setSalt(salt)
+
+    return this.migrateAndGetKey(password)
+    .then(() => {
+      return this.persistAllKeyrings()
+    })
+    .then(() => {
+      cb(null)
+    })
+    .catch((err) => {
+      cb(err)
+    })
+  }
+
+  createFirstKeyTree (password, cb) {
+    this.clearKeyrings()
+    this.addNewKeyring('HD Key Tree', {n: 1}, (err) => {
+      const firstKeyring = this.keyrings[0]
+      const accounts = firstKeyring.getAccounts()
+      const firstAccount = accounts[0]
+      const hexAccount = normalize(firstAccount)
+      const seedWords = firstKeyring.serialize().mnemonic
+      this.configManager.setSelectedAccount(firstAccount)
+      this.configManager.setSeedWords(seedWords)
+      autoFaucet(hexAccount)
+      this.setupAccounts(accounts)
+      this.persistAllKeyrings()
+      .then(() => {
+        cb(err, this.getState())
+      })
+      .catch((reason) => {
+        cb(reason)
+      })
+    })
+  }
+
+  placeSeedWords (cb) {
+    const firstKeyring = this.keyrings[0]
+    const seedWords = firstKeyring.serialize().mnemonic
+    this.configManager.setSeedWords(seedWords)
+  }
+
+  submitPassword (password, cb) {
+    this.migrateAndGetKey(password)
+    .then((key) => {
+      return this.unlockKeyrings(key)
+    })
+    .then((keyrings) => {
+      this.keyrings = keyrings
+      this.setupAccounts()
+      this.emit('update')
+      cb(null, this.getState())
+    })
+    .catch((err) => {
+      console.error(err)
+      cb(err)
+    })
+  }
+
+  loadKey (password) {
+    const salt = this.configManager.getSalt() || this.encryptor.generateSalt()
+    return this.encryptor.keyFromPassword(password + salt)
+    .then((key) => {
+      this.key = key
+      this.configManager.setSalt(salt)
+      return key
+    })
+  }
+
+  addNewKeyring (type, opts, cb) {
+    const Keyring = this.getKeyringClassForType(type)
+    const keyring = new Keyring(opts)
+    const accounts = keyring.getAccounts()
```

Comment:
not sure if accounts is available synchronously? should keyring be an event emitter and emit new keys that we then push to setup